src/machine/machine_avr.go

```diff
@@ -124,3 +124,47 @@ func (pwm PWM) Set(value uint16) {
 		panic("Invalid PWM pin")
 	}
 }
+
+// ADC on the Arduino
+const (
+	ADC0 = 0
+	ADC1 = 1
+	ADC2 = 2
+	ADC3 = 3
+	ADC4 = 4
+	ADC5 = 5
+)
+
+// InitADC initializes the registers needed for ADC.
+func InitADC() {
+	// set a2d prescaler so we are inside the desired 50-200 KHz range at 16MHz.
+	*avr.ADCSRA |= (avr.ADCSRA_ADPS2 | avr.ADCSRA_ADPS1 | avr.ADCSRA_ADPS0)
+
+	// enable a2d conversions
+	*avr.ADCSRA |= avr.ADCSRA_ADEN
+}
+
+// Configure configures a ADCPin to be able to be used to read data.
+func (a ADC) Configure() {
+	return // no pin specific setup on AVR machine.
+}
+
+// Get returns the current value of a ADC pin. The AVR will return a 10bit value ranging
+// from 0-1023.
+func (a ADC) Get() uint16 {
+	// set the analog reference (high two bits of ADMUX) and select the
+	// channel (low 4 bits), masked to only turn on one ADC at a time.
+	// this also sets ADLAR (left-adjust result) to 0 (the default).
+	*avr.ADMUX = avr.RegValue(avr.ADMUX_REFS0 | (a.Pin & 0x07))
```

Comment:
Can you enable ADLAR here, to get a result in the full 16-bit scale (with the low bits zero)?
Different devices vary in ADC precision but most often fall between 8 and 16 bits (nrf has 12 bits, for example) so the API would be most consistent with an ADC that uses the full 16-bit range.
And a small nitpick: the comment has full sentences but no capitals in front of them.

Reply:
I am confused by the request for ADLAR here. In either case the catse will be returned as a unit16 ranging from 0-1023 as it only has 10 bits available. 

The normal result for most ADC is a number from 0 to the max scale. Dividing that by the reference voltage gets the effective voltage for the ADC input. The result will fit into the uint16 so... not sure why the result to change the bit order of the internal registers.

Also regarding the comments: I am of the habit to use full sentences for the GoDoc comments, and smaller, uncapitalized for the case that there are inline comments. Is that a problem?

Reply:
Unless I'm misunderstanding the meaning of ADLAR here, I'm pretty sure when set it will return a value using the full 16-bit range but of course still with 10 bits of precision so the 6 lower bits are always 0. [This forum post](https://www.avrfreaks.net/comment/60470#comment-60470) explains it better than I can. Basically, when set, it will have an equivalent effect as shifting the result in ADCL/ADCH left by 6 bits. This is usually done if you only want to read the 8 high bits of the result (for 8 bits precision), but in this case it is an easy way to shift the result left by 6 without any additional effort.

My intention is the same as with ADC: make sure a program can use it without needing to know the range of the ADC. So for example, an Arduino has a range 0..1023 and nRF has a range of 0..4095. When both are scaled to 0..0xffff a user of the API can simply assume that 0x7fff means 2.5V with an input of 0..5V (for example).

Regarding comments: it's a very minor nit and probably personal style, so it's fine. I also write uncapitalized comments, but tend to use proper punctuation when it is a full sentence or more than one sentence.

Reply:
According to my reading of the AVR docs seems like ADLAR is intended for faster 8-bit conversions where the results are put in only one result register. Perhaps a constant per machine with the max value for ADC would be better to solve the scaling calculation.

Reply:
After reading that doc it seems unnecessary to  set ADLAR as we will have to read both registers in any case.

Reply:
ADLAR is indeed *meant* for faster 8-bit reads. But that doesn't mean it can't be used in a different way, namely to shift the output value (as both ADCH and ADCL are shifted).

Basically I'm arguing two things:
1. The output should be in the range 0..0xffff for easier portability. A possible way to implement that is to shift the result on AVR left by 6 bits, like so:
    ```go
    low := uint16(*avr.ADCL)
    high := uint16(*avr.ADCH)
    return (low <<6) | (high<<14))
    ```
2. The above result can be obtained in a more efficient way by setting the ADLAR bit, avoiding a shift operation.

Yes we can provide constants for the max value, and that would certainly be a good idea. But I think it is easier to write portable programs when all chips use the same 16-bit ADC range (but with only the available precision).

Reply:
> [...] where the results are put in only one result register.

That's not true: both registers are still set. The only difference is which bits are set in the result register. See the tables here: http://www.robotplatform.com/knowledge/ADC/adc_tutorial_3.html

Reply:
Seems like turning on the ADLAR bit and using the shifting code you suggested is not working for me as expected. Perhaps you would like to simply make that change yourself? I am not getting valid numbers like I was expecting.

Reply:
I'm getting the numbers as expected, so not quite sure where it went wrong with you. When connected to 3.3V I get 0x02a0 and 0x02a1 and when I set ADLAR I get 0xa800 and 0xa840 - exactly the shifted values (`val << 6`). The patch is https://github.com/aykevl/tinygo/commit/58b536619dcf4426de9ef1181877766e8c75615b.

Hint: you can print hexadecimal pointer-sized numbers by printing a number cast to uintptr.